Receive and verify AIK key

arturkow2000 opened the pull request.

arturkow2000 committed:
coap: implement block-wise transfers

This commit implements subset of RFC7959, bringing block transfer
support for incoming data. Currently can split data into at most 15
blocks of 1024 byte each, whichs allow to receive 15 KiB of data.

arturkow2000 committed:
coap: reduce block size to 512 bytes

When using 1024 bytes for block transfers network tends to lockup and
smoltcp fails to process incoming packets:

TRACE smoltcp::socket::udp     > #1:0.0.0.0:49152:169.254.0.8:5683: buffer to send 18 octets
 TRACE smoltcp::socket::udp     > #1:0.0.0.0:49152:0.0.0.0:49152: sending 18 octets
 DEBUG smoltcp::iface::ethernet > cannot process ingress packet: buffer space exhausted
 DEBUG smoltcp::iface::ethernet > packet dump follows:
EthernetII src=ea-96-91-87-26-8d dst=10-20-30-40-50-60 type=IPv4
\ IPv4 src=169.254.0.8 dst=169.254.0.1 proto=UDP
 \ UDP src=5683 dst=49152 len=1046
 TRACE smoltcp::socket::udp     > #1:0.0.0.0:49152:169.254.0.8:5683: buffer to send 18 octets
 TRACE smoltcp::socket::udp     > #1:0.0.0.0:49152:0.0.0.0:49152: sending 18 octets
 DEBUG smoltcp::iface::ethernet > cannot process ingress packet: buffer space exhausted
 DEBUG smoltcp::iface::ethernet > packet dump follows:
EthernetII src=ea-96-91-87-26-8d dst=10-20-30-40-50-60 type=IPv4
\ IPv4 src=169.254.0.8 dst=169.254.0.1 proto=UDP
 \ UDP src=5683 dst=49152 len=1046

Reducing block size to 512 allows to send 7 KiB of data, still it should
be more than enough.

arturkow2000 committed:
Begin working on certificate store implementation.

Implement basic certificate storage with certificate verification.
Currently no trust is implemented (root CAs can be stored but aren't
trusted) so certificate chain verification fails.

arturkow2000 committed:
tools: add tool for writing LFS

Add tools/lfs for manipulating contents of LittleFS-formatted flash.bin
file.

krystian-hebel committed:
Change CBOR major number for AIK modulus

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

arturkow2000 committed:
Obtain EK cert and AIK key from server.

Obtain EK cert and AIK key from server istead of keeping them hardcoded.
Add support for RSA AIK keys - most TPMs don't support ED25519 so it has
little to no use for us.

krystian-hebel committed:
src/client/proto.rs: add loaded_key_name field to AIK

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

arturkow2000 committed:
tools: add support for reading and writing LFS file attributes

arturkow2000 committed:
Support for marking certificates as trusted.

From now on CAs can be marked as trusted so chain verification should
work.

arturkow2000 committed:
client: don't check for signature size when using RSA

Trussed has a limit of MAX_SIGNATURE_LENGTH, but we are not using
Trussed for RSA so this limit does not apply. From now on signature
length is checked only for Ed25519.

arturkow2000 committed:
Check for unknown critical extensions, partially implement X.509v3 constraints.

During certificate verification constraints extension is respected. If
unknown critical extension is encountered verification fails as per
X.509 spec.

arturkow2000 committed:
tools/lfs: support installing certificates

Add a basic implementation of certificate installer. This is a simple
subcomand and it's neither a complete solution, actually it is a
temporary solution until Fobnail itself can manage its certificates.

arturkow2000 committed:
pal: increase flash size to 128 KiB

64 KiB is not enough and after saving a single root CA in certstore
Fobnail can't save metadata anymore. This is caused by
https://github.com/littlefs-project/littlefs/issues/645. 128 KiB is
enough for a few CAs and metadata.

arturkow2000 committed:
tools/lfs: don't quit on cert installation failure

Don't quit on first error, instead try to install
as many certificates as possible.

arturkow2000 committed:
certmgr: Fix verification of multi-level chain.

This is a one-line patch that fixes certificate chain verification when
there are intermediate CAs between the verified certificate and root
CA.

arturkow2000 committed:
Prevent panicking when Authority Key Id extension is present.

arturkow2000 committed:
trussed: use our own Trussed fork with custom patches

arturkow2000 committed:
Implement make credential operation.

arturkow2000 committed:
tpm: implement KDFa

Implement KDFa which is required for Credential Activation.
Implementation is based on https://github.com/google/go-tpm. Basic
testing have been done against original implementation.

arturkow2000 committed:
certmgr: make all DN fields optional

Don't make assumptions that any of DN field is present, require only
organization field in case of issuer, eventually this requirement should
also be relaxed.

arturkow2000 committed:
Implement AIK verification using Credential Activation.

arturkow2000 committed:
Harden AIK verification.

Previous approach for verifying AIK had was not secure, an attacker
could replace AIK because we were not checking whether received AIK
matches with it's name used for credential activation. Now, instead of
relying on AIK name provided from atteter we compute it ourselves. If
AIK has been tampered with, hashes will be different and credential
activation will fail.
Another advantage is that we can confirm that the received key has
proper attributes set, as required for AIK keys.

arturkow2000 committed:
Remove unneded definitions from client/proto.rs

We are receiving keys in TPM format, so we don't need these.

krystian-hebel commented on pal/pal_pc/src/trussed/store.rs:
```diff
@@ -63,9 +63,11 @@ impl littlefs2::driver::Storage for Flash {
+    // 64 KiB is not enough - LittleFS wastes much of available storage
+    // see https://github.com/littlefs-project/littlefs/issues/645
+    // increase size to 128 KiB
     const BLOCK_SIZE: usize = 4096;
```
Can we make it smaller in order to waste less space per directory?

krystian-hebel replied:
No, we can't, this is flash erase granularity.

krystian-hebel commented on src/certmgr/verify.rs:
```diff
@@ -0,0 +1,291 @@
+                    // While this works when there is a small number of certificates
+                    // it will fill memory when there are more.
+                    let potential_parents: Vec<X509Certificate> =
+                        self.iter_certificates(organization, trussed).collect();
```
Is it possible to return vector of file names (or paths if they are in different directories) instead and load full file contents inside `for` loop below?

arturkow2000 replied:
@krystian-hebel I modified `iter_certificates()` a bit so we can now call verify_internal while iterating.

krystian-hebel commented on src/client/mod.rs:
```diff
@@ -79,37 +85,142 @@ impl<'a> FobnailClient<'a> {
+                        // Rust won't allow us just move out data because doing
+                        // so would make state invalid.
+                        let mut aik = vec![];
+                        core::mem::swap(data, &mut aik);
```
I don't know about Rust's memory management, but shouldn't this be swapped back to initial `data` so it is properly freed?

krystian-hebel replied:
OK, I notices it is done in `handle_server_response()`. Maybe mention this in comments, here and in `handle_server_response()`?

arturkow2000 replied:
@krystian-hebel I mentioned it.

> I don't know about Rust's memory management, but shouldn't this be swapped back to initial data so it is properly freed?

No. When `aik` goes out of scope, Vec descrutctor will be called and data will freed normally.

krystian-hebel commented on src/client/mod.rs:
```diff
@@ -79,37 +85,142 @@ impl<'a> FobnailClient<'a> {
+                    request.set_path("/challenge");
+                    request.set_method(RequestType::Post);
+                    request.message.payload = encoded.to_vec();
+                    // TODO: should set Content-Type header to CBOR
```
```suggestion
                    request.message.set_content_format(coap_lite::ContentFormat::ApplicationCBOR);
```

This isn't checked on Attester side right now, but at some point we may depend on it.

arturkow2000 replied:
Fixed

arturkow2000 committed:
tools/lfs: add copy-from command

Add command from copying files from LittleFS to host.

arturkow2000 committed:
Add support for more X.509 extensions, do basic TCG compliance checks.

TCG mandates presence of some (critical) extensions in EK certificates.
This commit adds support for these extensions. Also it implements basic
checks to verify EK certificate compliance with TCG requirements.

arturkow2000 committed:
tools/lfs: support installing DER certificates, format code.

arturkow2000 committed:
Implement pathLenConstraint support.

krystian-hebel commented on src/client/state.rs:
```diff
@@ -12,20 +14,59 @@ pub enum State<'a> {
     // TODO: before requesting metadata we must request AIK (Attestation
     // Identity Key) key, receive it and verify.
     // This adds two new states: RequestAik and VerifyAik inserted between
     // InitDataReceived and RequestMetadata
```
Already done
```suggestion
```

krystian-hebel commented on src/client/tpm/kdf.rs:
```diff
@@ -0,0 +1,93 @@
+    bits: u32,
+) -> Vec<u8> {
+    // FIXME: Trussed has some support for KDF, but we cannot use it because of
+    // lack of RSA support (and possibly for other reasons two).
```
```suggestion
    // lack of RSA support (and possibly for other reasons too).
```

arturkow2000 committed:
Update Trussed.

Use Trussed version with modified MAX_MESSAGE_LENGTH. We need this
because some certificates are too big. Since we are using lfs tool for
uploading certificates we can write files bigger than that but we cannot
read them, Trussed will read only the first MAX_MESSAGE_LENGTH bytes
from file. In case of X.509 this may look as if certificate is corrupted
or the parser broken.

krystian-hebel commented on src/client/tpm/mod.rs:
```diff
@@ -0,0 +1,258 @@
+                    &[],
+                    (seed.len() * 8).try_into().unwrap(),
+                );
+                // panic!("AES key = {:#x?}", aes_key);
```
Remove

krystian-hebel commented on src/client/tpm/mu.rs:
```diff
@@ -0,0 +1,300 @@
+                ensure!(
+                    rsa_scheme.is_null() && rsa_scheme_details == 0x10,
+                    "Non-null scheme, this is not supported"
+                );
```
No, it is present, but it is `TPM2_ALG_NULL`. In that case its other fields (`keyBits`, `mode`, `details`) do not exist. This is easy to miss in TPM specification, it is hidden somewhere behind the definition of `+` symbol either prepended or appended to type name. Same is true for `scheme` - it is also `TPM2_ALG_NULL` so it has no `details`. In the current code what should be `symmetric.algorithm` is read as `scheme.scheme`, and what should be `scheme.scheme` is read as `scheme.detail`.

```suggestion
                let symmetric_algorithm_id = Algorithm::try_from_u8(decode!(cursor, data, u16))?;
                ensure!(
                    symmetric_algorithm_id.is_null(),
                    "Non-null symmetric algorithm, this is not supported"
                );
                
                let rsa_scheme = Algorithm::try_from_u8(decode!(cursor, data, u16))?;
                ensure!(
                    rsa_scheme.is_null(),
                    "Non-null scheme, this is not supported"
                );
```

arturkow2000 replied:
Fixed

arturkow2000 committed:
Set Content-Type to CBOR when sending CBOR-encoded objects

arturkow2000 committed:
certmgr: avoid loading all certificates at once

Apply a workaround so that instead of loading all certificates into
memory we can keep only one certificate loaded at time.

arturkow2000 committed:
tpm: fix TPM2B_PUBLIC parsing

macpijan commented:
@arturkow2000 @krystian-hebel What's left to close this one?

arturkow2000 commented:
@krystian-hebel Anything left?

krystian-hebel commented on src/client/tpm/mu.rs:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let buf_len: u16 = self.inner.len().try_into().expect("buffer to big for TPM");
```
```suggestion
        let buf_len: u16 = self.inner.len().try_into().expect("buffer too big for TPM");
```

arturkow2000 replied:
Fixed

krystian-hebel reviewed (commented):
One typo, otherwise LGTM

tomzy-0 commented:
Merged
